src/draw/lv_img_decoder.c

```diff
@@ -95,24 +95,13 @@ lv_res_t lv_img_decoder_get_info(const void * src, lv_img_header_t * header)
     return res;
 }
 
-/**
- * Open an image.
- * Try the created image decoder one by one. Once one is able to open the image that decoder is save in `dsc`
- * @param dsc describe a decoding session. Simply a pointer to an `lv_img_decoder_dsc_t` variable.
- * @param src the image source. Can be
- *  1) File name: E.g. "S:folder/img1.png" (The drivers needs to registered via `lv_fs_add_drv()`)
- *  2) Variable: Pointer to an `lv_img_dsc_t` variable
- *  3) Symbol: E.g. `LV_SYMBOL_OK`
- * @param color The color of the image with `LV_IMG_CF_ALPHA_...`
- * @return LV_RES_OK: opened the image. `dsc->img_data` and `dsc->header` are set.
- *         LV_RES_INV: none of the registered image decoders were able to open the image.
- */
-lv_res_t lv_img_decoder_open(lv_img_decoder_dsc_t * dsc, const void * src, lv_color_t color)
+lv_res_t lv_img_decoder_open(lv_img_decoder_dsc_t * dsc, const void * src, lv_color_t color, int32_t frame_id)
```

Comment:
with this design, each frame need call lv_img_decoder_open even all of them come from the same animation image. It isn't efficient since:

1. The same header need parse multiple time
2. The previous frame need decode multiple time if the format utilize the intra-frame compression 

And image widget need other important infomation: frame count and frame rate.

Reply:
Another approach is:

1. Add additional fields(frame_count, frame_rate...) to either lv_img_header_t or lv_img_decoder_dsc_t
2. Add frame_id or frame_number to lv_img_decoder_read_line instead of image_decoder_open

Frame cache perspective:

1. LVGL image decoder cache work on the file level
2. image decoder can do frame level cache internall if needed.

Reply:
I've reread our discussion from [here](https://github.com/lvgl/lvgl/issues/2090#issuecomment-786017827) and now I see that there was a little bit of misunderstanding in it.

So let's consider gif decoding as a real-life use case. If I understood correctly your suggestion is:
1. The `lv_gif` widget knows the timing of the frames and it starts a timer with the time to the next frame
2. When the timer expires a `frame_id` variable in `lv_gif` is incremented,  the image is invalidated, and a new timer is started.
3. When the image is redrawn the new `frame_id` is passed to the drawer.
4. The cache sees that this file is already open and some parts might (e.g. color tables) might be already loaded and it decided to use it regardless of the difference in the `frame_id`.
5. The image decoder opens the image again and sees the difference in `frame_id` so it internally calculates the new frame, saves it, and returns it.
6. LVGL draws the new frame.

It sounds reasonable but I see things to discuss:
- If there are 2 animations with the same source running asynchronously the active frame of both can't be cached. Or only with an ugly logic in the decoder that caches the last N requested frame.  
- In point 5 (above): now LVGL doesn't call "decoder_open" again on cache hit. That's the purpose of caching to avoid opening (decompressing) the image again and again. If the image is already open, calling an other open on it would be strange anyway. However, we could have a `update_cb` on the decoder that is called even if there was a cache hit. Usually, it does nothing, but with gif it can see that now there is a new frame to load. 
- `lv_gif` needs to contain a simple gif decoder that can extract the duration of the frames. And we need a "real" gif decoder too that actually decodes the gif. It's not a critical issue but IMO it'd be better to keep the image decoding details in one place.

Reply:
> 5. The image decoder opens the image again and sees the difference in `frame_id` so it internally calculates the new frame, saves it, and returns it.

For simple stateless decoder, it can open the image again. But for the advanced state decoder, it may keep the file in the open state until we call lv_img_decoder_close_f_t.

> 6. LVGL draws the new frame.
> 
> It sounds reasonable but I see things to discuss:
> 
> * If there are 2 animations with the same source running asynchronously the active frame of both can't be cached. Or only with an ugly logic in the decoder that caches the last N requested frame.

Yes, but decoder can implement the full feature cache by hash table if they want.

> * In point 5 (above): now LVGL doesn't call "decoder_open" again on cache hit. That's the purpose of caching to avoid opening (decompressing) the image again and again. If the image is already open, calling an other open on it would be strange anyway. However, we could have a `update_cb` on the decoder that is called even if there was a cache hit. Usually, it does nothing, but with gif it can see that now there is a new frame to load.

After more thinking, we can do the frame level cache in the common code too with my suggestion by putting the cache info into _lv_img_decoder_dsc(basically we need extend _lv_img_decoder_dsc::img_data to an array/hash).

> * `lv_gif` needs to contain a simple gif decoder that can extract the duration of the frames. And we need a "real" gif decoder too that actually decodes the gif. It's not a critical issue but IMO it'd be better to keep the image decoding details in one place.

Reply:
BTW, most animation decoder is design to open once, decode frame one by one and then close. It's better to let lv_img_decoder_t follow this convention to simplify the integration.

Reply:
>> The image decoder opens the image again and sees the difference in frame_id so it internally calculates the new frame, saves it, and returns it.
> For simple stateless decoder, it can open the image again. But for the advanced state decoder, it may keep the file in the open state until we call lv_img_decoder_close_f_t.

True, but with a stateless decoder we can not leverage the cache.

>> If there are 2 animations with the same source running asynchronously the active frame of both can't be cached. Or only with an ugly logic in the decoder that caches the last N requested frame.
> Yes, but decoder can implement the full feature cache by hash table if they want.

True but the decoder needs to know how many frames to buffer which depends on the number of visible `lv_gif`s. If the cache happens on a frame level, we will have exactly as many cached frames as many visible `lv_gif`s.

> After more thinking, we can do the frame level cache in the common code too with my suggestion by putting the cache info into _lv_img_decoder_dsc(basically we need extend _lv_img_decoder_dsc::img_data to an array/hash).

The `img_data` would be indexed by the `frame_id`, right? If so what if the img_cahce sees that the image is open, but that frame is missing? Call `decoder_open` again? Or the `update_cb` I've mentioned above. 

> BTW, most animation decoder is design to open once, decode frame one by one and then close. It's better to let lv_img_decoder_t follow this convention to simplify the integration.

It also means that the frames can be decoded only sequentially. And we can get (easily) frame 1, then frame 36, then frame 2, then frame 37. 

Does it mean that you agree to keep frame-level cache?

Reply:
> > > The image decoder opens the image again and sees the difference in frame_id so it internally calculates the new frame, saves it, and returns it.
> 
> > For simple stateless decoder, it can open the image again. But for the advanced state decoder, it may keep the file in the open state until we call lv_img_decoder_close_f_t.
> 
> True, but with a stateless decoder we can not leverage the cache.
> 
> > > If there are 2 animations with the same source running asynchronously the active frame of both can't be cached. Or only with an ugly logic in the decoder that caches the last N requested frame.
> 
> > Yes, but decoder can implement the full feature cache by hash table if they want.
> 
> True but the decoder needs to know how many frames to buffer which depends on the number of visible `lv_gif`s. If the cache happens on a frame level, we will have exactly as many cached frames as many visible `lv_gif`s.
> 

The cache can use LRU algo to discard the longest unused item.

> > After more thinking, we can do the frame level cache in the common code too with my suggestion by putting the cache info into _lv_img_decoder_dsc(basically we need extend _lv_img_decoder_dsc::img_data to an array/hash).
> 
> The `img_data` would be indexed by the `frame_id`, right? If so what if the img_cahce sees that the image is open, but that frame is missing? Call `decoder_open` again? Or the `update_cb` I've mentioned above.
> 

No, it should call lv_img_decoder_read_line_f_t for frame missing. decoder_open is used for file missing. 

> > BTW, most animation decoder is design to open once, decode frame one by one and then close. It's better to let lv_img_decoder_t follow this convention to simplify the integration.
> 
> It also means that the frames can be decoded only sequentially. And we can get (easily) frame 1, then frame 36, then frame 2, then frame 37.
> 

Yes, we can if frame_id add to lv_img_decoder_read_line_f_t. Of course, the decoder need support seeking to the right place.

> Does it mean that you agree to keep frame-level cache?

Yes, I think it may be useful.